components/support/rc_crypto/src/pbkdf2.rs

```diff
@@ -0,0 +1,75 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use crate::error::*;
+use nss::pbkdf2::pbkdf2_key_derive;
+pub use nss::pbkdf2::HashAlgorithm;
+/// Extend passwords using pbkdf2, based on the following [rfc](https://www.ietf.org/rfc/rfc2898.txt) it runs the NSS implementation
+/// # Arguments
+///
+///  * `passphrase` - The password to stretch
+///  * `salt` - A salt to use in the generation process
+///  * `iterations` - The number of iterations the hashing algorithm will run on each section of the key
+///  * `hash_algorithm` - The hash algorithm to use
+///  * `out` - The slice the algorithm will populate
+///
+/// # Examples
+///
+/// ```
+/// use rc_crypto::pbkdf2;
+/// let mut password = b"password".to_vec();
+/// let mut salt = b"salt".to_vec();
+/// let mut out = vec![0u8; 32];
+/// pbkdf2::derive(&mut password, &mut salt, 2, pbkdf2::HashAlgorithm::SHA256, &mut out).unwrap(); // Oh oh should handle the error!
+/// assert_eq!(hex::encode(out), "ae4d0c95af6b46d32d0adff928f06dd02a303f8ef3c251dfd6e2d85a95474c43");
+//
+///```
+///
+/// # Errors
+///
+/// Could possibly return an error if the HMAC algorithm fails, or if the NSS algorithm returns an error
+pub fn derive(
+    passphrase: &mut [u8],
+    salt: &mut [u8],
+    iterations: u32,
+    hash_algorithm: HashAlgorithm,
+    out: &mut [u8],
+) -> Result<()> {
+    pbkdf2_key_derive(passphrase, salt, iterations, hash_algorithm, out)?;
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    #[test]
+    fn test_generate() {
+        let expected = "120fb6cffcf8b32c43e7225256c4f837a86548c92ccc35480805987cb70be17b";
+        let mut out = vec![0u8; 32];
+        let mut password = b"password".to_vec();
+        let mut salt = b"salt".to_vec();
+        derive(&mut password, &mut salt, 1, HashAlgorithm::SHA256, &mut out).unwrap();
+        assert_eq!(expected, hex::encode(out));
+    }
+
+    #[test]
+    fn test_longer_key() {
+        let expected = "120fb6cffcf8b32c43e7225256c4f837a86548c92ccc35480805987cb70be17b4dbf3a2f3dad3377264bb7b8e8330d4efc7451418617dabef683735361cdc18c";
+        let mut password = b"password".to_vec();
+        let mut salt = b"salt".to_vec();
+        let mut out = vec![0u8; 64];
+        derive(&mut password, &mut salt, 1, HashAlgorithm::SHA256, &mut out).unwrap();
+        assert_eq!(expected, hex::encode(out));
+    }
+
+    #[test]
+    fn test_more_iterations() {
+        let expected = "ae4d0c95af6b46d32d0adff928f06dd02a303f8ef3c251dfd6e2d85a95474c43";
+        let mut password = b"password".to_vec();
+        let mut salt = b"salt".to_vec();
+        let mut out = vec![0u8; 32];
+        derive(&mut password, &mut salt, 2, HashAlgorithm::SHA256, &mut out).unwrap();
+        assert_eq!(expected, hex::encode(out));
+    }
+}
```

Comment:
This seems to be working as expected, but I'd like to see more coverage of edge-cases in these tests, particularly around the things that might be going wrong when we transit from Rust into C. For example: do we properly handle embedded nulls in the password, and the salt? can we generate keys that aren't a full multiple of the hash blocksize? do we cleanly reject if requesting more than `(2^32 - 1) * hLen` bytes of output, assuming that NSS correctly fails in this case?

I think that `hkdf.rs` has a good sample of some of the things worth worrying about, but I also took the liberty of generating what I think would be some worthwhile extra test vectors, using the python `cryptography` library:

```
PBKDF2-SHA256(b'password', b'salt', 1, 32) = b'120fb6cffcf8b32c43e7225256c4f837a86548c92ccc35480805987cb70be17b'
PBKDF2-SHA256(b'password', b'salt', 1, 64) = b'120fb6cffcf8b32c43e7225256c4f837a86548c92ccc35480805987cb70be17b4dbf3a2f3dad3377264bb7b8e8330d4efc7451418617dabef683735361cdc18c'
PBKDF2-SHA256(b'password', b'salt', 2, 32) = b'ae4d0c95af6b46d32d0adff928f06dd02a303f8ef3c251dfd6e2d85a95474c43'
PBKDF2-SHA256(b'password', b'salt', 3, 33) = b'ad35240ac683febfaf3cd49d845473fbbbaa2437f5f82d5a415ae00ac76c6bfccf'
PBKDF2-SHA256(b'passw\x00rd', b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00', 5, 13) = b'e25d526987819f966e324faa4a'
PBKDF2-SHA256(b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00', b'salt', 2, 32) = b'62384466264daadc4144018c6bd864648272b34da8980d31521ffcce92ae003b'
PBKDF2-SHA256(b'password', b'salt', 1, 0) = b''
PBKDF2-SHA256(b'', b'salt', 1, 32) = b'f135c27993baf98773c5cdb40a5706ce6a345cde61b000a67858650cd6a324d7'
PBKDF2-SHA256(b'password', b'', 1, 32) = b'c1232f10f62715fda06ae7c0a2037ca19b33cf103b727ba56d870c11f290a2ab'
PBKDF2-SHA256(b'password', b'salt', 0, 32) = b'120fb6cffcf8b32c43e7225256c4f837a86548c92ccc35480805987cb70be17b'                                                                                                     
```

Reply:
Thank you for those!! I added the tests, but one thing I did notice, is that the NSS algorithm rejects 0 iterations and 0 length keyLength (out.len() ). I confirmed that via Crypto.subtle. But otherwise added the tests you mentioned and a few others to tests correct rejections.

Reply:
> NSS algorithm rejects 0 iterations and 0 length keyLength

Great :-)

TBH I was kind of surprised that the python implementation allowed these, but it's worth having tests to document the specific behaviour in these cases either way.